File: acmp/Путь.cpp
//https://acmp.ru/index.asp?main=task&id_task=127

#include <iostream>
#include <vector>
#include <map>
#include <queue>

using namespace std;

vector<int> BFS(vector<vector<int>> graph, int size, int start)
{
    vector<int> dist(size, -1);
    queue<int> suspended;

    dist[start] = 0;

    suspended.push(start);

    while (!suspended.empty())
    {
        auto current = suspended.front();
        suspended.pop();

        for (int adjecent : graph[current])
        {
            if (dist[adjecent] == -1)
            {
                dist[adjecent] = dist[current] + 1;
                suspended.push(adjecent);
            }
        }
    }

    return dist;
}

int main()
{
    int n;
    cin >> n;
    vector<vector<int>> graph(n);

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            int has;
            cin >> has;

            if (has == 1)
            {
                graph[i].push_back(j);
                graph[j].push_back(i);
            }
        }
    }

    int u, v;
    cin >> u >> v;
    u--; v--;
    auto dist = BFS(graph, n, u);

    cout << dist[v];
}

File: acmp/Путь.cs
//https://acmp.ru/index.asp?main=task&id_task=127

using System;
using System.Collections.Generic;
using System.Linq;

public class Program
{
    public static int[] BFS(List<int>[] connections, int startingVertex)
    {
        int[] routes = Enumerable.Repeat(-1, connections.Length).ToArray();
        int[] parents = Enumerable.Repeat(-1, connections.Length).ToArray();
        Queue<int> suspended = new Queue<int>();

        suspended.Enqueue(startingVertex);
        routes[startingVertex] = 0;

        while (suspended.Count > 0)
        {
            var currentVertex = suspended.Dequeue();

            if (routes[currentVertex] == -1)
            {
                routes[currentVertex] = routes[parents[currentVertex]] + 1;
            }

            foreach (var connection in connections[currentVertex])
            {
                if (parents[connection] == -1)
                {
                    parents[connection] = currentVertex;
                    suspended.Enqueue(connection);
                }
            }
        }

        return routes;
    }
    public static void Main(string[] args)
    {
        var n = int.Parse(Console.ReadLine());
        var connections = new List<int>[n];

        for (int i = 0; i < n; i++)
        {
            connections[i] = new List<int>();
        }

        for (int i = 0; i < n; i++)
        {
            var input = Console.ReadLine().Split(new string[] { " " }, StringSplitOptions.RemoveEmptyEntries);

            for (int j = 0; j < input.Length; j++)
            {
                if (input[j] == "1")
                {
                    connections[i].Add(j);
                    connections[j].Add(i);
                }
            }
        }

        var target = Console.ReadLine().Split(new string[] { " " }, StringSplitOptions.RemoveEmptyEntries);
        var firstTargeted = int.Parse(target[0]) - 1;
        var secondTargeted = int.Parse(target[1]) - 1;
        var routes = BFS(connections, firstTargeted);

        Console.WriteLine(routes[secondTargeted]);
    }
}